### ISSUE TITLE
ConcatenatingAudioSource with lots of children takes a very long time to load

### ISSUE THREAD

**Issue by smkhalsa** [ryanheise/just_audio#294]
**Which API doesn't behave as documented, and how does it misbehave?**

Creating a ConcatenatingAudioSource with lots (~1000) of children takes a very long time (>20 seconds).

In my application, users can have playlists with an arbitrary number of items.

**Minimal reproduction project**

**To Reproduce (i.e. user steps, not code)**

```dart
final player = AudioPlayer();
final songs = /// 1000+ sources
await player.setAudioSource(
  ConcatenatingAudioSource(children: songs),
);
```

**Error messages**

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS: MacOS

**Smartphone (please complete the following information):**
 - Device: iPhone 12 Pro Max Simulator

**Flutter SDK version**
```
[✓] Flutter (Channel beta, 1.25.0-8.3.pre, on Mac OS X 10.15.7 19H114 darwin-x64, locale en-US)
[✓] Android toolchain - develop for Android devices (Android SDK version 29.0.2)
[✓] Xcode - develop for iOS and macOS
[✓] Chrome - develop for the web
[✓] Android Studio (version 4.1)
[✓] VS Code (version 1.52.1)
[✓] Connected device (2 available)

• No issues found!
```

**Additional context**

**Comment by ryanheise** [comment]
Hi @smkhalsa unfortunately I will need you to fill in the sections as instructed. In the section called "User steps, not code", you actually copied and pasted code. This should be in the previous section "Minimal reproduction project" which you have left empty, but which should contain a link to a git repository that I can clone and test.

**Comment by ryanheise** [comment]
This issue affects both iOS and macOS. I have a reproduction case on #206 which has helped me to reproduce this issue.

The solution is to rewrite the `enqueueFrom` method so that it enqueues only the next few items, and whenever the player advances to the next item, we should add another item onto the end. This behaviour should be the default, but it should also be tied to the `ConcatenatingAudioSource.useLazyPreparation` option.

This is a tricky area of the code to work on since it can easily break things, but it is an important feature to add, and as long as it undergoes sufficient testing, it should be worth doing.

I've marked this issue as "fixing" which means it is on my priority list. That said, there are some other things that were already on my priority list, such as the visualizer, and the null-safe release of audio_service (which will be based on the one-isolate model).

**Comment by suragch** [comment]
@ryanheise The `enqueFrom` you are talking about is [here](https://github.com/ryanheise/just_audio/blob/77c6f9e03a9e4f25642e0077895c721f78991b38/just_audio/darwin/Classes/AudioPlayer.m)? So that's not something we could handle on plugin user side, right? 

I was noticing similar behavior with 43 items in the playlist. My UI freezes while they are loading on macOS and iOS. On the iOS simulator the freeze lasted 30 seconds. Web and Android are fine.

**Comment by ryanheise** [comment]
That's correct. The solution involves improving the `enqueueFrom` so that it can enqueue just a few items ahead instead of the whole list, when the `useLazyPreparation` option is set to true. That involves modifying the Objective C code in the plugin.

As for a workaround, I can think of 2 things to try:

1. The easiest thing to try (although I don't know whether it will have any effect) is to play with the new buffering options available in the `dev` branch, which you can pass into the constructor of `AudioPlayer`. With these options (in particular `preferredForwardBufferDuration`), you may be able to limit the size of the forward buffer and thus reduce the time spent loading. The problem with this is that iOS isn't guaranteed to follow your preferred buffer parameters, and even if it did, it still may need to load a little bit of each track.
2. Forego just_audio's playlist management and implement the lazy loading within your own app. So let's say you have a list of 43 items, you could create a `ConcatenatingAudioSource` with just the first 3 items in it. Then when playback approaches the 3rd item, you can dynamically insert more items onto the end of it. The drawback is that you can't simply use just_audio's state to tell you what's in your entire playlist, because it'll only know about the sublist of items you've added, so you'll need to maintain your own full playlist state outside of just_audio.

**Comment by suragch** [comment]
I forgot about the possibility of adding items to an existing playlist. That's a good idea. Thank you!

**Comment by YaarPatandarAA** [comment]
Just to add onto this, I have only noticed this issue on MacOS release/debug and only on iOS Simulator. This issue is not present, at least for me, on Physical iOS device whether release or debug.

**Comment by FallenChromium** [comment]
As far as I understand, the question that I have is heavily related to this issue (though not exactly relevant, I use Android and it isn't a "bug" per say, it's just that retrieving 40+ URLs is time-costly with the API I use)

I am not sure if I can inject a function to lazy load the track URL in `AudioSource` objects, so that the URL retrieval (and also caching) will kick in only when the track is about to begin playing. Did I get it correctly that it's now not an option, unless I make an "external" playlist state and control just_audio according to that playlist?

**Comment by ryanheise** [comment]
`ConcatenatingAudioSource.useLazyPreparation` is already implemented on Android, it just needs to be implemented on iOS/macOS.

**Comment by FallenChromium** [comment]
As far as I understood, `useLazyPreparation` is used to delay caching, but not the retrieval of children in `ConcatenatingAudioSource`, so, if I use API which, for example, returns download links which are valid only for 1 minute, this is not an option for me. I was initially supposing that after setting this bool, the `children` themselves will be resolved lazily. So, it's not related to the issue? How can I achieve this kind of effect using just_audio then?

**Comment by ryanheise** [comment]
This issue is about iOS/macOS. If you have found an Android bug or have an Android feature request, it will be appreciated to open a separate issue for that.

**Comment by ertgrulll** [comment]
Any progress about the issue? I'm developing an app for windows/mac and player is freezing for 4-5 seconds when i add 60 song.

**Comment by ryanheise** [comment]
The current AVQueuePlayer will work fine, but the enqueue logic should be rewritten to enqueue a configurable number of items, and at the same time, the gapless looping approach should be adapted to this. It's going to take a while to develop and test.

However, you can work around this by doing the same sort of queue management in dart.  That is, only add to the concatenating audio source as many children as you want to actually load, then lazily add more children just before they're needed.

**Comment by pro100svitlo** [comment]
Any update on this issue?
For me UI locking even with 100 items (around 7-10 sec).
Quite sad experience...

**Comment by ryanheise** [comment]
This will always be problematic to some degree because of Dart's single threaded model. Fortunately there are ways to deal with this which I have already mentioned (in my edit at the bottom of the very first post).

**Comment by mhutshow** [comment]
Hi. I have tried a lot to play with your workaround. I was programatically adding the track in the list. But unfortunately it didn't work for me.

It's a humble request to you. Please work for a fix. We love your work, You made it amazing. <3

Hope this will be fixed soon.

**Comment by ryanheise** [comment]
I know that people are using that workaround successfully (lazily adding items to the playlist just-in-time so that you don't try to load them all at once). I suggest you keep trying to implement that workaround because it is going to be much harder to build multi-threaded behaviour into the plugin and you'll be waiting much longer.

**Comment by mhutshow** [comment]
@ryanheise [ryanheise](https://github.com/ryanheise) : Can you provide us a code example that how can I lazily load items? 

```
I tried for (var i = 0; i < 100000; i++) {
  children.add(...);
 await Future.delayed(Duration.zero);
}
```

**Comment by pro100svitlo** [comment]
For me lazy loading didn't work.
But i implemented another suggestion: instead adding whole list at once I added 3 items: previous, current and next.
Then if user click next, or previous, i add one new playlist item at the start of at the end of the list.

To be honest - it looks ugly. But when it's covered with the tests - somehow you can trust it

**Comment by mhutshow** [comment]
I think lazy loading will work. Because if it listen to the list that I am adding new item then it will work.

But in my case the problem is when it loads, it loads with initial items. Then I keep adding the file by a for loop. But it don't see the changes (new files). After playing the initial files it stops.

I think if we get an example code from @ryanheise  then it will be easy to implement the feature. 

Let's wait with a hope from @ryanheise .

**Comment by ryanheise** [comment]
@pro100svitlo 's approach should work, it's a treadmill with always 3 items.

I'm not going to implement it here @mhutshow but @pro100svitlo has given the pseudocode in his previous comment. If the treadmill contains 3 items [0,1,2], then each time the current index changes to 0 or 2, you can do the treadmill thing that @pro100svitlo describes well.

**Comment by mhutshow** [comment]
@ryanheise @pro100svitlo : My case is little bit different . There is no next or previous button. the tracks are 5-6 seconds long. A track is divided into 30-260 pieces and I have to play it one by one. If I add item to the children after it starts playing the initial 3, the newly added item can not be played. Sadly, It stops after playing the initial 3.

**Comment by pro100svitlo** [comment]
The point is that you need to add one more each time next clicked or current ended.
In this way you should always have one item ahead

**Comment by mirkancal** [comment]
Is it possible now with the `LockCachingAudioSource`?

**Comment by ryanheise** [comment]
@mirkancal What are you referring to by the word "it"? Do you mean one of the workarounds or pseudocode mentioned above, and if so, which one?

**Comment by mirkancal** [comment]
@ryanheise sorry I was reading lot of issues, I meant asking it under Playing a remote file and cache the file at the same time
https://github.com/ryanheise/just_audio/issues/47

**Comment by MrVipinVijayan** [comment]
use preload: false
It will fix the ui freezing issue.

**Comment by mhutshow** [comment]
Believe me. This only happens in the iOS simulator. I have a 100K users production app. And no issue comes regarding freezing. 
So just ignore it during development in the simulator. In production/real device no issue will come.

**Comment by MrVipinVijayan** [comment]
Good to hear that, but unfortunately for me, it happens in debug and release mode, both in simulator and device.

**Comment by mirkancal** [comment]
Just removing the `await` keyword helped in my case.

```
// ignore: unawaited_futures
 _player.setAudioSource(_concatenatingAudioSource);
```

As soon as I get my sounds from the backend, I send a custom action to the handler.

 ```
 @override
  Future<dynamic> customAction(String name,
      [Map<String, dynamic>? extras]) async {
    switch (name) {
      case 'load':
        final urls = extras!['urls'] as List<String>;
        if (!kIsWeb) {
          cachedAudios =
              urls.map((e) => LockCachingAudioSource(Uri.parse(e))).toList();
        } else {
          cachedAudios =
              urls.map((e) => ProgressiveAudioSource(Uri.parse(e))).toList();
        }

        _concatenatingAudioSource = ConcatenatingAudioSource(
          children: cachedAudios,
        );
        // ignore: unawaited_futures
        _player.setAudioSource(_concatenatingAudioSource);
        break;
    }
  }
```

**Comment by ryanheise** [comment]
I am recently considering implementing the treadmill approach within the Dart side of the plugin, and that would involve removing the existing Android implementation which is currently implemented entirely on the native side. If this is implemented as a treadmill in Dart, then it should work on all platforms consistently, but there is also a chance that the Android behaviour will not be exactly the same as before (although I can't predict yet whether that would cause problems).

**Comment by suragch** [comment]
That sounds like a great solution to me. It would remove complexity on the native side and solve the only major issue i have with the plugin (slow ios load times).

**Comment by ryanheise** [comment]
Hmm, while this idea is attractive at first glance, I just spent a little time thinking about how I would implement this approach and I don't think it would be compatible with the way just_audio_background works because just_audio_background needs to be aware of the whole queue at all times in order to broadcast the queue, not merely the part of the queue that is on the top surface of the treadmill. So the platform will need to know about the complete list of children. Another issue is that such an implementation may introduce a lot of complexity and hence bugs.

I guess I will be implementing the iOS treadmill in native code after all.

**Comment by ryanheise** [comment]
I have just implemented an iOS/macOS treadmill. It is available for testing on the `feature/treadmill` branch.

Currently it is hard coded to be always on because it is actually difficult with the current design to emulate the Android behaviour (that may influence a future API change sot hat lazy loading is a property of the player rather than of the playlist).

Also, the treadmill is hardcoded to hold two items at a time. Ideally the length of the treadmill would dynamically adapt based on the shortness of the items. For example, for short items, you'd want to preload more items onto the treadmill to ensure gapless playback. For longer items, the current and next item would be enough.

**Comment by suragch** [comment]
Wow, that was fast! Thank you!

**Comment by ryanheise** [comment]
It took me 1.5 years to find a quick solution :-) But even still I dreaded changing the iOS queuing logic since it was complex to write the first version and it's important not to introduce bugs. Hopefully we can see through testing whether it is solid or has any issues.

**Comment by austinried** [comment]
I found this thread because I have a similar issue with slow load times for large playlists, but in my case it's more related to things I need to do while constructing the playlist (async check a cache & potentially perform a network request per song).  I'm wondering, and I haven't put enough thought into this yet to put together a full feature request, but @ryanheise would you consider a solution to this that involves something like a "childBuilder" method for ConcatenatingAudioSource that implements this treadmill on the dart side, and also allows us to do those arbitrary just-in-time preparations on the library user end?

**Comment by ryanheise** [comment]
As stated above, that approach will not work due to the way the platform interface works.

However, with the current API it is completely possible for an app to implement a simple treadmill of its own with all of its app-specific requirements. You can listen to `player.currentIndexStream` and whenever it changes, you can trigger your own logic to add the next item to the `ConcatenatingAudioSource` according to your own business logic.

**Comment by austinried** [comment]
Ah understood, I will go that route then.  It's a bit of a shame because it means I'll also need to re-implement shuffle/loop and the events that go with those, but that's not a huge deal I suppose.

**Comment by ryanheise** [comment]
Is anyone able to report on their experience with the solution implemented in `feature/treadmill`? As mentioned, I am reluctant to merge changes to this core code unless people find that it is actually working well for them. Once it's proven to be an effective solution, I can merge it into an official release.

**Comment by Yesterday17** [comment]
The `currentPlaying` index stream seems wrong when I test it last week.

**Comment by ryanheise** [comment]
Does this happen for you on `example_playlist.dart`?

**Comment by burhanaksendir** [comment]
> I have just implemented an iOS/macOS treadmill. It is available for testing on the `feature/treadmill` branch.
> 
> Currently it is hard coded to be always on because it is actually difficult with the current design to emulate the Android behaviour (that may influence a future API change sot hat lazy loading is a property of the player rather than of the playlist).
> 
> Also, the treadmill is hardcoded to hold two items at a time. Ideally the length of the treadmill would dynamically adapt based on the shortness of the items. For example, for short items, you'd want to preload more items onto the treadmill to ensure gapless playback. For longer items, the current and next item would be enough.

Wow, that was fast! Thank you! With that, problem solved.

**Comment by burhanaksendir** [comment]
> Is anyone able to report on their experience with the solution implemented in `feature/treadmill`? As mentioned, I am reluctant to merge changes to this core code unless people find that it is actually working well for them. Once it's proven to be an effective solution, I can merge it into an official release.

Can you merge this into an official version? Thanks.

**Comment by ryanheise** [comment]
Thank you for testing it, but I need to hear from people who have been using it for a while to know whether it is actually stable before I merge it.

**Comment by burhanaksendir** [comment]
> Thank you for testing it, but I need to hear from people who have been using it for a while to know whether it is actually stable before I merge it.

 Any chance to update feature/treadmill? In feature/treadmill, I cannot access some features in the official release version (eg positionDiscontinuityStream etc..).

**Comment by ryanheise** [comment]
Updated.

**Comment by xweng1016** [comment]
treadmill probably should be merged, been on this branch for like half a year now

**Comment by ryanheise** [comment]
@xweng1016 and yet you never provided any testing feedback on it. So no, it will not be merged yet.

(Earlier comment:)

> Thank you for testing it, but I need to hear from people who have been using it for a while to know whether it is actually stable before I merge it.

**Comment by mz5210** [comment]
> 

我测试了502个音乐，批量加入播放列表中，它运行良好，没有卡顿，也许您应该把这个改动合并到master
<img width="791" alt="image" src="https://user-images.githubusercontent.com/29273353/223049964-a194082f-8daa-41b4-8163-aa5160907899.png">

**Comment by mz5210** [comment]
@ryanheise

**Comment by ryanheise** [comment]
Hi @mz5210 Please write your comments in English, thank you.

**Comment by mz5210** [comment]
@ryanheise I tested 502 pieces of music, adding them to the playlist in bulk, it worked fine, no stuck, maybe you should incorporate this change into master

**Comment by burhanaksendir** [comment]
Hi @ryanheise,
Any chance to update feature/treadmill?

**Comment by ryanheise** [comment]
Just updated.

**Comment by zhahouming** [comment]
@ryanheise works fine!
![image](https://user-images.githubusercontent.com/16803650/226240174-507dca69-e74e-466e-8820-f1a3fd3200b3.png)

**Comment by UnicornsOnLSD** [comment]
First impressions on `feature/treadmill` - my weird issue with gapless sometimes being broken is fixed! I'll use it on my own device for a bit to see how stable it is :)

**Comment by jagpreetsethi** [comment]
@ryanheise Would this already address a scenario where audio media needs to be loaded from an HTTP URL which becomes available after making an additional request ? 

Something like below:
Items loaded in the playlist is <List of HTML URLs>
When actually playing a music, the item from the above list is HTTP requested, which returns an MP3 audio file to play. 

If feature/treadmill doesn't address this, what approach would you recommend ?

Thanks in advance.

**Comment by ryanheise** [comment]
#777 ?

**Comment by LouisAldorio** [comment]
so I can load 2000 audio now directly to the player without any freezing issue ?

**Comment by ryanheise** [comment]
@LouisAldorio Try it and let me know how it goes. The treadmill branch has been put out there to be tested in order to receive feedback.

**Comment by LouisAldorio** [comment]
Branch feature/treadmill right ? , tomorrow I will try it let you know , but just to confirm I am not using the wrong branch , feature/treadmill right , sir ryan

**Comment by ryanheise** [comment]
> but just to confirm I am not using the wrong branch , feature/treadmill right

Well, if you find any other treadmill branch, please let me know about it!

**Comment by LouisAldorio** [comment]
everything works well, 917 tracks, added and instantly played lol. that was fast

**Comment by hunterwilhelm** [comment]
Even though mine instantly started playing, it took about 30 seconds to add 1000 tracks in the newest version of feature/treadmill on iOS with all the flags enabled. I see a huge improvement, but it is still not satisfactory for shuffle mode to work properly when starting the app.

**Comment by ryanheise** [comment]
@hunterwilhelm Can you just clarify, so it started playing BEFORE it added the tracks? In which case, what precisely do you mean by adding the tracks? Do you mean the time it took to appear in the UI?

**Comment by hunterwilhelm** [comment]
@ryanheise Here is how I have it in the code right now.
1. My app initializes
2. Download the URLs of the tracks and start loading them one by one using `.add` instead of `.addAll` because `.addAll` is still freezing the UI.
3. After the first add, the UI looks fine,
    * but shuffle mode will be very wrong because it won't take the whole playlist into account. Rather, it would only shuffle the songs that have been added so far, so when you hit next, you have an 0% chance of getting the last track in the playlist before it's done. Not a true shuffle - yet.
5. The user presses play, and it instantly plays because the beginning of the playlist has already loaded.
6. Then after 30 seconds after step 2, the tracks have all been added.
7. At this point, theoretically, I can use shuffle again, but it's too late. I would like to have a working shuffle without having to wait.

Right now I am making my own implementation of the dynamically loading so it works on iOS. I am loading the tracks into the playlist one by one when needed. But if the user starts in the middle of the playlist, they need to go backward. So, I am going to have to write a complicated function to map the indices of the queues (shuffled and chronological) to the playlist indices (for precaching purposes) and vice versa.

**Comment by ryanheise** [comment]
In the case of `.addAll` or passing all children into the constructor, there are several points in the code that cause the slowdown. The `treadmill` branch addresses only one of them, which is to prevent the native player on iOS from trying to load the entire playlist all at once. But there are still 2 other bottlenecks. One is passing a large playlist over the flutter method channel all at once, and the other is purely on the dart side when processing long lists. Dart is not multi-threaded, so any time we do a for loop over a long list, that is potentially going to lock up the UI.

If you do have a very long playlist, yes, you do have to just avoid passing very long lists into `ConcatenatingAudioSource` and instead use it more dynamically so that you maintain your own Dart list containing your entire playlist, but you use `ConcatenatingAudioSource` to only ever contain the next 3 items to play. When you reach the end of the current track, you remove it from the front and add a new track onto the back so that you always have 3 items queued up for gapless playback.

The long term future direction of the plugin should be to shift the internal implementation somewhat in this direction. There are complications to that, so there is no short term quick fix, apart from using `ConcatenatingAudioSource` within your app in a more dynamic way as described above.

**Comment by jagpreetsethi** [comment]
@ryanheise  Is there any example code available that I may refer that follows your suggested approach to fill ConcatenatingAudioSource dynamically by 3 items only? Thanks in advance!

**Comment by hunterwilhelm** [comment]
@ryanheise thanks for that quick response! That makes sense. Someday I'll be knowledgeable enough on the topic to be able to assist with the amazing work you do!

@jagpreetsethi Once I finish my implemention of what Ryan is talking about, I'll post it here. It will include shuffling.

**Comment by burhanaksendir** [comment]
Hi @ryanheise 

I've been trying to include the just_audio package in my Flutter project using its Git repository as the source. However, I've been encountering an issue when specifying the Git reference in the pubspec.yaml file.

Here's how I'm adding the package to my pubspec.yaml file:

```
just_audio:
  git:
    url: https://github.com/ryanheise/just_audio.git
    ref: feature/treadmill

```

However, every time I attempt to fetch the package, I get an error indicating that the pubspec.yaml file cannot be found in the specified Git reference:

> Could not find a file named "pubspec.yaml" in https://github.com/ryanheise/just_audio.git 02f74512e51c40ba68fe4c898d570c0c4ae906d5.

The reference I'm using is: feature/treadmill, and I've verified that it exists in the repository: https://github.com/ryanheise/just_audio/tree/feature/treadmill.

I've tried cleaning the cache, updating Flutter packages, and even cloning the repository locally to test, but the issue persists.

Is there something I might be missing or any suggestions you can provide to help me resolve this? I greatly appreciate your assistance in this matter.

Thank you!

**Comment by ryanheise** [comment]
It's because just_audio isn't in the root directory of this repo, it's in the `just_audio` subdirectory, so you need to specify that with `path: just_audio`. So:

```yaml
just_audio:
  git:
    url: https://github.com/ryanheise/just_audio.git
    ref: feature/treadmill
    path: just_audio
```

**Comment by Eldar2021** [comment]
For me this `feature/treadmill` works good. I would ask you to merge it. Thank you for the nice package and your support.

**Comment by hunterwilhelm** [comment]
@jagpreetsethi I implemented what @ryanheise was suggesting. However, I wasn't able to figure out shuffling while preloading without having any visual glitches or refreshes when switching between shuffle mode and non-shuffle mode. So I left it out. If anyone can modify it to add it, I would be very grateful.

<details>
  <summary>
    The Implementation:
  </summary>

```dart
import 'dart:async';
import 'dart:math' show max, min;

import 'package:audio_service/audio_service.dart';
import 'package:just_audio/just_audio.dart';
import 'package:quiver/iterables.dart' show range;
import 'package:rxdart/rxdart.dart';

Future<AudioHandler> initAudioService() async {
  return await AudioService.init(
    builder: () => MyAudioHandler(),
    config: const AudioServiceConfig(
      androidNotificationChannelId: 'com.example.app.audio',
      androidNotificationChannelName: 'Example App',
      androidNotificationOngoing: true,
      androidStopForegroundOnPause: true,
    ),
  );
}

class PlaylistManager {
  final _player = AudioPlayer();
  final _subPlaylist = ConcatenatingAudioSource(
    children: [],
    useLazyPreparation: false, // I want it to pre-cache everything I put in this playlist.
  );

  /// A copy of the media items so we can access them later
  final List<MediaItem> _subPlaylistItems = [];

  /// The full list of media items to pull from
  final List<MediaItem> _fullPlaylist = [];

  /// Used for keeping track of how many tracks are in front of the current index of the player.
  int _playlistFrontLoadingCount = 0;

  /// Change this if you want the index to not start at the beginning
  static const kDefaultIndex = 0;

  /// Keeps track of the current index of the _fullPlaylist
  int _currentFullPlaylistIndex = kDefaultIndex;

  /// Used for cleaning up the subscriptions when disposed
  final List<StreamSubscription> _streamSubscriptions = [];

  /// Used for preventing race conditions in the _lazyLoadPlaylist
  bool _updatePlayerRunning = false;

  /// Used for preventing race conditions in the _lazyLoadPlaylist
  bool _updatePlayerWasCalledWhileRunning = false;

  /// Used for not skipping too many times if the [_subPlaylist] isn't done loading yet.
  int _directionOfSkipTo = 0;

  // PUBLIC VARS / GETTERS

  /// Notifies the system of the current media item
  final BehaviorSubject<MediaItem> mediaItem = BehaviorSubject();

  /// Use this to get the live data of what the state is of the player
  ///
  /// Don't perform actions like skip on this object because there is extra logic
  /// attached to the functions in this class. For example, use [seekToNext] instead of [player.seekToNext]
  AudioPlayer get player => _player;

  /// The index of the next item to be played
  int get nextIndex => _getRelativeIndex(1);

  /// The index of the previous item in play order
  int get previousIndex => _getRelativeIndex(-1);

  /// The index of the current index.
  int get currentIndex => _currentFullPlaylistIndex;

  /// How many tracks next and previous are loaded into ConcatenatingAudioSource
  /// * Ex. 3 would look like [`prev3`, `prev2`, `prev1`, `current`, `next1`, `next2`, `next3`]
  final int preloadPaddingCount;

  PlaylistManager({this.preloadPaddingCount = 3}) {
    _attachAudioSourceToPlayer();
    _listenForCurrentSongIndexChanges();
  }

  /// Used for loading the tracks. This will reset the current index and position.
  setQueue(List<MediaItem> mediaItems) async {
    _fullPlaylist.clear();
    _fullPlaylist.addAll(mediaItems);
    _subPlaylist.clear();
    _subPlaylistItems.clear();
    _lazyLoadPlaylist();
  }

  /// Use this instead of [player.seekToNext]
  Future<void> seekToNext() async {
    _directionOfSkipTo = 1;
    await _player.seekToNext();
  }

  /// Use this instead of [player.seekToPrevious]
  Future<void> seekToPrevious() async {
    _directionOfSkipTo = -1;
    await _player.seekToPrevious();
  }

  Future<void> _attachAudioSourceToPlayer() async {
    try {
      await _player.setAudioSource(_subPlaylist);
    } catch (e) {
      print("Error: $e");
    }
  }

  void _listenForCurrentSongIndexChanges() {
    int? previousIndex = _player.currentIndex;
    _streamSubscriptions.add(_player.currentIndexStream.listen((index) {
      _updateMediaItem();
      _lazyLoadPlaylist();

      final previousIndex_ = previousIndex;
      previousIndex = index;
      if (previousIndex_ == null || index == null) return;
      final delta = index - previousIndex_;

      if (delta.sign == _directionOfSkipTo.sign) {
        _currentFullPlaylistIndex += delta;
        _playlistFrontLoadingCount += delta;
      }
    }));
  }

  int _getRelativeIndex(int offset) {
    return max(0, min(_currentFullPlaylistIndex + offset, _fullPlaylist.length - 1));
  }

  _updateMediaItem() {
    if (_subPlaylistItems.isEmpty) return;
    final playerIndex = _player.currentIndex;
    if (playerIndex == null) return;

    final newMediaItem = _subPlaylistItems[playerIndex];
    mediaItem.add(newMediaItem);
  }

  _lazyLoadPlaylist() async {
    // prevent race conditions
    if (_updatePlayerRunning) {
      _updatePlayerWasCalledWhileRunning = true;
      return;
    }
    _updatePlayerRunning = true;

    final currentIndex_ = _currentFullPlaylistIndex;
    final playerIndex = _player.currentIndex ?? 0;

    // Pad/pre-cache the ending of the playlist
    final currentNextPadding = max(0, _subPlaylist.length - playerIndex);
    var nextCountToAdd = preloadPaddingCount - currentNextPadding + 1;
    if (nextCountToAdd > 0 && _fullPlaylist.isNotEmpty) {
      for (final iNum in range(nextCountToAdd)) {
        var mediaItem = _fullPlaylist[iNum.toInt() + currentIndex_];
        await _subPlaylist.add(_createAudioSource(mediaItem));
        _subPlaylistItems.add(mediaItem);
        await Future.microtask(() {});
      }
    }

    // Pad/pre-cache the beginning of the playlist
    final currentPreviousPadding = _player.currentIndex ?? 0;
    final previousCountToAdd = preloadPaddingCount - currentPreviousPadding;
    if (previousCountToAdd > 0) {
      for (int i = 1; i <= previousCountToAdd; i++) {
        var index = currentIndex_ - currentPreviousPadding - _playlistFrontLoadingCount - i;
        if (index < 0 || _fullPlaylist.length <= index) continue;
        var mediaItem = _fullPlaylist[index];
        final future = _subPlaylist.insert(0, _createAudioSource(mediaItem));
        _subPlaylistItems.insert(0, mediaItem);
        _playlistFrontLoadingCount++;
        await future;
        await Future.microtask(() {});
      }
    }

    _updateMediaItem();

    // prevent race conditions
    _updatePlayerRunning = false;
    if (_updatePlayerWasCalledWhileRunning) {
      _updatePlayerWasCalledWhileRunning = false;
      Future.microtask(() {
        _lazyLoadPlaylist();
      });
    }
  }

  UriAudioSource _createAudioSource(MediaItem mediaItem) {
    return AudioSource.uri(
      Uri.parse(mediaItem.extras!['url'] as String),
      tag: mediaItem,
    );
  }

  Future<void> dispose() async {
    for (final subscription in _streamSubscriptions) {
      await subscription.cancel();
    }
    _streamSubscriptions.clear();
    return _player.dispose();
  }
}

class MyAudioHandler extends BaseAudioHandler {
  final _playlistManager = PlaylistManager();
  final List<StreamSubscription> _streamSubscriptions = [];

  MyAudioHandler() {
    _notifyAudioHandlerAboutPlaybackEvents();
    _listenForDurationChanges();
    _listenForCurrentSongIndexChanges();
  }

  void _notifyAudioHandlerAboutPlaybackEvents() {
    _streamSubscriptions.add(_playlistManager.player.playbackEventStream.listen((PlaybackEvent event) {
      final playing = _playlistManager.player.playing;
      playbackState.add(playbackState.value.copyWith(
        controls: [
          MediaControl.skipToPrevious,
          if (playing) MediaControl.pause else MediaControl.play,
          MediaControl.stop,
          MediaControl.skipToNext,
        ],
        systemActions: const {
          MediaAction.seek,
        },
        androidCompactActionIndices: const [0, 1, 3],
        processingState: const {
          ProcessingState.idle: AudioProcessingState.idle,
          ProcessingState.loading: AudioProcessingState.loading,
          ProcessingState.buffering: AudioProcessingState.buffering,
          ProcessingState.ready: AudioProcessingState.ready,
          ProcessingState.completed: AudioProcessingState.completed,
        }[_playlistManager.player.processingState]!,
        repeatMode: const {
          LoopMode.off: AudioServiceRepeatMode.none,
          LoopMode.one: AudioServiceRepeatMode.one,
          LoopMode.all: AudioServiceRepeatMode.all,
        }[_playlistManager.player.loopMode]!,
        shuffleMode: (_playlistManager.player.shuffleModeEnabled) ? AudioServiceShuffleMode.all : AudioServiceShuffleMode.none,
        playing: playing,
        updatePosition: _playlistManager.player.position,
        bufferedPosition: _playlistManager.player.bufferedPosition,
        speed: _playlistManager.player.speed,
        queueIndex: event.currentIndex,
      ));
    }));
  }

  void _listenForDurationChanges() {
    _streamSubscriptions.add(_playlistManager.player.durationStream.listen((duration) {
      final oldMediaItem = _playlistManager.mediaItem.valueOrNull;
      if (oldMediaItem == null) return;
      final newMediaItem = oldMediaItem.copyWith(duration: duration);
      mediaItem.add(newMediaItem);
    }));
  }

  void _listenForCurrentSongIndexChanges() {
    _streamSubscriptions.add(_playlistManager.mediaItem.listen((newMediaItem) {
      final newMediaItemWithDuration = newMediaItem.copyWith(duration: _playlistManager.player.duration);
      mediaItem.add(newMediaItemWithDuration);
    }));
  }

  @override
  Future<void> addQueueItems(List<MediaItem> mediaItems) async {
    throw UnimplementedError("addQueueItems");
  }

  @override
  Future<void> addQueueItem(MediaItem mediaItem) async {
    throw UnimplementedError("addQueueItem");
  }

  @override
  Future<void> updateQueue(List<MediaItem> mediaItems) async {
    // notify system
    final newQueue = mediaItems;
    queue.add(newQueue);
    _playlistManager.setQueue(mediaItems);
  }

  @override
  Future<void> removeQueueItemAt(int index) async {
    throw UnimplementedError("removeQueueItemAt");
  }

  @override
  Future<void> play() => _playlistManager.player.play();

  @override
  Future<void> pause() => _playlistManager.player.pause();

  @override
  Future<void> seek(Duration position) => _playlistManager.player.seek(position);

  @override
  Future<void> skipToQueueItem(int index) async {
    throw UnimplementedError("skipToQueueItem");
  }

  @override
  Future<void> skipToNext() async {
    _playlistManager.seekToNext();
  }

  @override
  Future<void> skipToPrevious() async {
    if (_playlistManager.player.position.inSeconds > 5 || _playlistManager.currentIndex == 0) {
      return _playlistManager.player.seek(Duration.zero);
    } else {
      return _playlistManager.seekToPrevious();
    }
  }

  @override
  Future<void> setRepeatMode(AudioServiceRepeatMode repeatMode) async {
    throw UnimplementedError("setRepeatMode");
  }

  @override
  Future<void> setShuffleMode(AudioServiceShuffleMode shuffleMode) async {
    throw UnimplementedError("setShuffleMode");
  }

  @override
  Future<void> customAction(String name, [Map<String, dynamic>? extras]) async {
    if (name == 'dispose') {
      for (final subscription in _streamSubscriptions) {
        await subscription.cancel();
      }
      _streamSubscriptions.clear();

      await _playlistManager.dispose();
      super.stop();
    }
  }

  @override
  Future<void> onTaskRemoved() async {
    stop();
    super.onTaskRemoved();
  }

  @override
  Future<void> stop() async {
    await _playlistManager.player.stop();
    return super.stop();
  }
}
```

</details>

**Comment by hunterwilhelm** [comment]
@ryanheise Is there a way to disable preloading/pre-caching entirely for `ConcatenatingAudioSource`? I don't see an option to turn it off on iOS.

**Comment by ryanheise** [comment]
No, that is in fact why you are working around it by only adding a limited number of items to the playlist. You need to have at least the current and the next item in the playlist in order to get gapless playback. If you want to get more sophisticated, you could listen to the current position, and when it reaches 15-20 seconds before reaching the end of the track, that's when you can add the next item to the playlist, so that it then starts buffering it just in time.

**Comment by adamhaqiem** [comment]
> No, that is in fact why you are working around it by only adding a limited number of items to the playlist. You need to have at least the current and the next item in the playlist in order to get gapless playback. If you want to get more sophisticated, you could listen to the current position, and when it reaches 15-20 seconds before reaching the end of the track, that's when you can add the next item to the playlist, so that it then starts buffering it just in time.

Would it be a problem if I don't remove the previous item and just keep adding for as much as I want?

**Comment by ryanheise** [comment]
If you keep adding without ever removing, you will eventually end up with a very large list that could again slow down communication between Dart and iOS.

**Comment by Eldar2021** [comment]
Can you please update branch(`feature/treadmill`)?

**Comment by ryanheise** [comment]
What specifically would you like me to update about it?

**Comment by burhanaksendir** [comment]
Dear @ryanheise,
Firstly, congratulations on this fantastic package. I've encountered an issue – I've created an MP3 podcast playlist with no spaces at the beginning of the files, yet there's a disruptive effect during transitions. Is it possible to add a very brief gap at each episode change, or is there a way to smooth out the transitions?

**Comment by ryanheise** [comment]
Hi @burhanaksendir , this issue concerns slow loading of long playlists on iOS only. I would like to maintain separate issues for separate issues, so I'll mark both of our comments as off topic.

**Comment by ulutashus** [comment]
@ryanheise Is there any reason, known issue to not merge feature/treadmill branch? First impression, it works just fine for me. I am planning to publish it to our production if there is no known issue. Btw Is this branch changes affecting Android too?

**Comment by ryanheise** [comment]
I think one more thing I would like to do is make this behaviour sensitive to the `useLazyPreparation` parameter, or even add a separate parameter to specifically control the iOS/macOS specific behaviour. In this case, perhaps the parameter would control the size of the treadmill. I welcome any feedback on this.

**Comment by ryanheise** [comment]
> or even add a separate parameter to specifically control the iOS/macOS specific behaviour. In this case, perhaps the parameter would control the size of the treadmill. I welcome any feedback on this.

The only reason to want to specify the size of the treadmill is for short queue items where items are so short that having two items on the treadmill is not enough of a buffer to ensure that we will not hit the end of the buffer and stall playback. However, I think a boolean option such as `useLazyPreparation` might suffice after all. Apps that are using "really" short items where this is a consideration probably aren't the intended use case for lazy loading anyway, and they have two options:

1. Switch off the option. The entire playlist may load fast enough anyway if the items are really short.
2. Implement their own treadmill in Dart using the mutating methods in `ConcatenatingAudioSource`.

So unless there are any objections, I may just tie this to `useLazyPreparation` with a treadmill size of 2.

**Comment by ryanheise** [comment]
Update: I have made substantial changes to the `feature/treadmill` branch in order to support the `useLazyPreparation` option, along with a couple of other bug fixes/enhancements.

Please let me know if this update breaks anything in your apps before I officially merge it into the next release.

**Comment by MohammadElKhatib** [comment]
> Update: I have made substantial changes to the `feature/treadmill` branch in order to support the `useLazyPreparation` option, along with a couple of other bug fixes/enhancements.
> 
> Please let me know if this update breaks anything in your apps before I officially merge it into the next release.

Working great, I am using it to load 114 files with large files such as 2 hours length with no issues

**Comment by bekchan** [comment]
> Update: I have made substantial changes to the `feature/treadmill` branch in order to support the `useLazyPreparation` option, along with a couple of other bug fixes/enhancements.
> 
> Please let me know if this update breaks anything in your apps before I officially merge it into the next release.

Works awesome with 1000+ ConcatenatingAudioSources in playlist.

**Comment by ryanheise** [comment]
Thanks for the testing results @MohammadElKhatib and @bekchan .

Have either of you had any experience with how this interacts with shuffle mode, loop modes, or edge cases such as where there are only 1 or 0 items in the playlist?

**Comment by MohammadElKhatib** [comment]
> Thanks for the testing results @MohammadElKhatib and @bekchan .
> 
> Have either of you had any experience with how this interacts with shuffle mode, loop modes, or edge cases such as where there are only 1 or 0 items in the playlist?

Well I believe it is working so well as I mentioned previously, first it did not work well but maybe it was caching the default branch and once I cleared cache and re-run the application, it is working awesome

I am using it for Quran application where there is multiple reciters so basically 114 files with around 15 reciters and what i am doing is loading all the files and replace it once changing the reciters also I am listening to the sound progress updates in order to highlight text in parallel with sound.

thanks for your efforts waiting the merge with main branch :-)

**Comment by ryanheise** [comment]
Thanks, @MohammadElKhatib . Are you using either/both of shuffle and loop modes, or is that not a pertinent feature in your app?

**Comment by MohammadElKhatib** [comment]
> Thanks, @MohammadElKhatib . Are you using either/both of shuffle and loop modes, or is that not a pertinent feature in your app?

I used the loop options but I faced one issue I am not sure if this is the default behavior, what happened is that I used a button to toggle the loop and repeat. I have 3 options repeat one, repeat all and no repeat basically repeat one is repeating the current sound normally, repeat all is going through all the files and repeat from first index after reaching the last index while the no repeat is keep going forward till the last sound and stop there.

while what I wanted is to only play one sound and stop so I am not sure if this is a bug or default behavior or I am missing something here.

Note: I am also using background service and syncing text and progress alltogether.

below screenshot of my app

![IMG_8010](https://github.com/ryanheise/just_audio/assets/5835334/1ac20080-24f5-4e2b-b800-c191f8683742)

**Comment by ryanheise** [comment]
> while what I wanted is to only play one sound and stop so I am not sure if this is a bug or default behavior or I am missing something here.

If you have a `ConcatenatingAudioSource`, it is the normal behaviour to auto-advance to the next child after completing the current one. There is another feature request to make that configurable, but until then you could listen to the `PositionDiscontinuity` event type and when it is `autoAdvance`, you can handle that event by pausing. The playlist example contains some code that shows how to listen for this event, since it uses it to show a toast whenever auto-advancing. Since this is a workaround until the aforementioned feature is implemented, it is not going to be perfect, though, because the time between when you receive the event and when you invoke `pause()` and wait for it to actually pause won't be zero, and so there is a chance that if the next audio doesn't have any silent padding at the start, you might hear a few milliseconds of the next audio before it pauses.

**Comment by MohammadElKhatib** [comment]
> > while what I wanted is to only play one sound and stop so I am not sure if this is a bug or default behavior or I am missing something here.
> 
> If you have a `ConcatenatingAudioSource`, it is the normal behaviour to auto-advance to the next child after completing the current one. There is another feature request to make that configurable, but until then you could listen to the `PositionDiscontinuity` event type and when it is `autoAdvance`, you can handle that event by pausing. The playlist example contains some code that shows how to listen for this event, since it uses it to show a toast whenever auto-advancing. Since this is a workaround until the aforementioned feature is implemented, it is not going to be perfect, though, because the time between when you receive the event and when you invoke `pause()` and wait for it to actually pause won't be zero, and so there is a chance that if the next audio doesn't have any silent padding at the start, you might hear a few milliseconds of the next audio before it pauses.

Honestly, the current situation is acceptable in my app, I will keep following your updates.

Thanks again for your great work.

**Comment by bekchan** [comment]
> Have either of you had any experience with how this interacts with shuffle mode, loop modes, or edge cases such as where there are only 1 or 0 items in the playlist?

Shuffle and loop modes works great. Also tested reorderable playlist with +1000 audio, still good.
Thank you for this great package.

**Comment by ryanheise** [comment]
I have just merged and published the `feature/treadmill` branch. Thank you very much to those who helped to test it. This issue will remain open since there are 2 other sources of lag that can be introduced by long lists. One is that the Dart iterators don't yield, which may be an issue on very long lists, and the other is (or was?) possibly the lag from passing a huge message over the method channels all at once. I have to test whether this is still an issue. But of course in the meantime, apps can work around this by creating an empty `ConcatenatingAudioSource` and incrementally building it up rather than initialising it all in one go.

**Comment by milesegan** [comment]
I'm still seeing this long delay here on iOS despite making sure `useLazyPreparation` is enabled. I've also tried breaking up longer lists and building up the `ConcatenatingAudioSource` source in chunks instead of all at once. 

Are other people still seeing this issue or is this something specific I'm doing wrong in my own code?

**Comment by InShad7** [comment]
In iOS using ConcatenatingAudioSource getting stuck on when 1 or 2 audios played.

**Comment by InShad7** [comment]
any solutions?

**Comment by InShad7** [comment]
> > Update: I have made substantial changes to the `feature/treadmill` branch in order to support the `useLazyPreparation` option, along with a couple of other bug fixes/enhancements.
> > Please let me know if this update breaks anything in your apps before I officially merge it into the next release.
> 
> Working great, I am using it to load 114 files with large files such as 2 hours length with no issues



> > Update: I have made substantial changes to the `feature/treadmill` branch in order to support the `useLazyPreparation` option, along with a couple of other bug fixes/enhancements.
> > Please let me know if this update breaks anything in your apps before I officially merge it into the next release.
> 
> Works awesome with 1000+ ConcatenatingAudioSources in playlist.





can you please suggest the way u did !
iam getting stuck on iOS after 1 or 2 audio plays using concatetaingAudioSource

**Comment by sinhpn92** [comment]
I have the same issue on iOS. I just used the AudioSource list to `setAudioSources`. Did you solve your issue @InShad7 ?

**Comment by InShad7** [comment]
from the package side the issue still exist,
i created a new logic according to the my app functionality.
but some time, it happening.
